<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Intro_to_MiCBuS â€¢ MiCBuS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Intro_to_MiCBuS">
<meta property="og:description" content="MiCBuS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MiCBuS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Intro_to_MiCBuS.html">Intro_to_MiCBuS</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Intro_to_MiCBuS</h1>
            
      
      
      <div class="hidden name"><code>Intro_to_MiCBuS.Rmd</code></div>

    </div>

    
    
<p>Marker Gene Mining for Unknown Cell Types Using Bulk and Single Cell
RNA-Seq Data</p>
<p><img src="Fig1.png" width="90%" style="display: block; margin: auto;"></p>
<p>A framework of MiCBuS to identify marker genes for unknown cell
types. The unknown cell types, highlighted in orange, are present in
bulk RNA-seq data but remain undetected in incomplete scRNA-seq data.
MiCBuS takes both bulk and single-cell RNA-seq data as input and follows
three major steps. MiCBuS estimates cell type proportions in bulk
RNA-seq data using scRNA-seq as a reference. It then proceeds to
generate Dirichlet-pseudo-bulk RNA-seq samples. By comparing these
simulated Dirichlet-pseudo-bulk RNA-seq samples with the original bulk
RNA-seq data, MiCBuS can identify pseudo marker genes of unknown cell
types, referred to as psMarker.</p>
<p>This is an example of how to use <strong>MiCBuS</strong>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load bulk RNA-seq data (raw count table)</span></span>
<span><span class="va">bulkdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"bulkDat.csv"</span>, package <span class="op">=</span> <span class="st">"MiCBuS"</span><span class="op">)</span>,</span>
<span>  row.names <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="va">bulkdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">bulkdat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load cell type RNA-seq data (sc or bulk)</span></span>
<span><span class="va">sc_eset_incomplete</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sc_eset_incomplete.rds"</span>, package <span class="op">=</span> <span class="st">"MiCBuS"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run MiCBuS() function to get the DE table for pseudo marker gene for unknown cell types.</span></span>
<span><span class="va">DEpseudo.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MiCBuS.html">MiCBuS</a></span><span class="op">(</span>bulkdat<span class="op">=</span><span class="va">bulkdat</span>,sc_eset<span class="op">=</span><span class="va">sc_eset_incomplete</span>,m<span class="op">=</span><span class="fl">20</span>, seed<span class="op">=</span><span class="fl">1234</span><span class="op">)</span></span></code></pre></div>
<p>To evaluate how well the marker genes for unknown cell types are
detected in the psMarker gene set, we use the scMarker genes identified
from complete scRNA-seq data. In the performance evaluation, we used the
Jaccard index, a commonly used metric to measure how similar two sets of
genes are in terms of their shared elements. It is calculated by
dividing the size of the intersection of the sets (the number of genes
they have in common) by the size of the union of the sets (the total
number of unique genes in both sets). The resulting value ranges from 0
to 1, where 0 indicates no overlap, and 1 indicates complete
overlap.</p>
<p><strong>Table 1. Evaluation of Performance of MiCBuS Using Jaccard
Index</strong></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select top pseudo-bulk marker genes from MiCBuS DE results</span></span>
<span><span class="va">psMarkerTop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/psMarker_Top.html">psMarker_Top</a></span><span class="op">(</span>DEpseudo.df <span class="op">=</span> <span class="va">DEpseudo.df</span>, ntop <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load marker genes (Top 200 for each cell type) identified from complete scRNA-seq data </span></span>
<span><span class="co"># Only applicable when the gene profile of unknown cell type is given</span></span>
<span><span class="va">scMarkersTop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"scMarkersTop.csv"</span>, package <span class="op">=</span> <span class="st">"MiCBuS"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare psMarker genes with marker genes identified from complete scRNA-seq data</span></span>
<span><span class="co"># Helper: compute intersection count + Jaccard index</span></span>
<span><span class="va">calc_jaccard</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">set1</span>, <span class="va">set2</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">inter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">set1</span>, <span class="va">set2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">jac</span>   <span class="op">&lt;-</span> <span class="va">inter</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">union</a></span><span class="op">(</span><span class="va">set1</span>, <span class="va">set2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>n_intersect <span class="op">=</span> <span class="va">inter</span>, jaccard <span class="op">=</span> <span class="va">jac</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Identify the missing cluster(s)</span></span>
<span><span class="va">mis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">scMarkersTop</span><span class="op">$</span><span class="va">cluster</span>, <span class="fu">pData</span><span class="op">(</span><span class="va">sc_eset_incomplete</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Split marker genes into mis and other(unique)</span></span>
<span><span class="va">mis_genes</span>   <span class="op">&lt;-</span> <span class="va">scMarkersTop</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">scMarkersTop</span><span class="op">$</span><span class="va">cluster</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">mis</span><span class="op">]</span></span>
<span><span class="va">other_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">scMarkersTop</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="op">!</span><span class="va">scMarkersTop</span><span class="op">$</span><span class="va">cluster</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">mis</span><span class="op">]</span>, <span class="va">mis_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate Jaccard for each comparison</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>  Unknown_Cell_Type    <span class="op">=</span> <span class="fu">calc_jaccard</span><span class="op">(</span><span class="va">psMarkerTop</span>, <span class="va">mis_genes</span><span class="op">)</span>,</span>
<span>  Other_unique   <span class="op">=</span> <span class="fu">calc_jaccard</span><span class="op">(</span><span class="va">psMarkerTop</span>, <span class="va">other_genes</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Prepare a clean data frame</span></span>
<span><span class="va">jaccard_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  scMarker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>,</span>
<span>  Number_of_intersected_genes <span class="op">=</span> <span class="va">res</span><span class="op">[</span>, <span class="st">"n_intersect"</span><span class="op">]</span>,</span>
<span>  Jaccard_index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>, <span class="st">"jaccard"</span><span class="op">]</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  row.names <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># output table</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="va">jaccard_df</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">scMarker</th>
<th align="right">Number_of_intersected_genes</th>
<th align="right">Jaccard_index</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Unknown_Cell_Type</td>
<td align="right">48</td>
<td align="right">0.136</td>
</tr>
<tr class="even">
<td align="left">Other_unique</td>
<td align="right">0</td>
<td align="right">0.000</td>
</tr>
</tbody>
</table>
<p>The comparison of psMarkers and scMarkers can also be shown in this
Venn diagram.This figure includes three gene sets, <strong>ps</strong>
(psMarker genes of unknown cell types identified by MiCBuS) and scMarker
(<strong>mis</strong>: marker genes of the unknown cell types identified
using unmasked scRNA-seq data; <strong>other</strong>: marker genes of
the other known cell types identified using unmasked scRNA-seq
data.)</p>
<p><img src="Intro_to_MiCBuS_files/figure-html/Visualization-1.png" width="288" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>R version 4.3.1 (2023-06-16 ucrt) Platform: x86_64-w64-mingw32/x64
(64-bit) Running under: Windows 10 x64 (build 19045)</p>
<p>Matrix products: default</p>
<p>locale: [1] LC_COLLATE=English_United States.utf8 [2]
LC_CTYPE=English_United States.utf8<br>
[3] LC_MONETARY=English_United States.utf8 [4] LC_NUMERIC=C<br>
[5] LC_TIME=English_United States.utf8</p>
<p>time zone: America/New_York tzcode source: internal</p>
<p>attached base packages: [1] stats graphics grDevices utils datasets
methods base</p>
<p>other attached packages: [1] Biobase_2.60.0 BiocGenerics_0.46.0
knitr_1.46<br>
[4] ggVennDiagram_1.5.2 ggplot2_3.5.1 dplyr_1.1.4<br>
[7] MiCBuS_0.1.0</p>
<p>loaded via a namespace (and not attached): [1] RColorBrewer_1.1-3
rstudioapi_0.16.0<br>
[3] jsonlite_1.8.8 magrittr_2.0.3<br>
[5] spatstat.utils_3.2-0 farver_2.1.1<br>
[7] rmarkdown_2.26 zlibbioc_1.46.0<br>
[9] fs_1.6.4 ragg_1.3.0<br>
[11] vctrs_0.6.5.9000 ROCR_1.0-11<br>
[13] memoise_2.0.1 spatstat.explore_3.2-7<br>
[15] RCurl_1.98-1.14 S4Arrays_1.0.6<br>
[17] htmltools_0.5.8.1 sass_0.4.9<br>
[19] sctransform_0.4.1 parallelly_1.37.1<br>
[21] KernSmooth_2.23-22 bslib_0.7.0<br>
[23] htmlwidgets_1.6.4 desc_1.4.3<br>
[25] ica_1.0-3 plyr_1.8.9<br>
[27] plotly_4.10.4 zoo_1.8-12<br>
[29] cachem_1.0.8 alabama_2023.1.0<br>
[31] igraph_2.0.3 mime_0.12<br>
[33] lifecycle_1.0.4 pkgconfig_2.0.3<br>
[35] Matrix_1.6-4 R6_2.5.1<br>
[37] fastmap_1.1.1 GenomeInfoDbData_1.2.10<br>
[39] MatrixGenerics_1.12.3 fitdistrplus_1.1-11<br>
[41] future_1.33.2 shiny_1.8.1.1<br>
[43] digest_0.6.35 numDeriv_2016.8-1.1<br>
[45] colorspace_2.1-0 S4Vectors_0.38.2<br>
[47] patchwork_1.2.0 DESeq2_1.40.2<br>
[49] Seurat_5.0.3 tensor_1.5<br>
[51] RSpectra_0.16-1 irlba_2.3.5.1<br>
[53] GenomicRanges_1.52.1 textshaping_0.3.7<br>
[55] labeling_0.4.3 progressr_0.14.0<br>
[57] fansi_1.0.6 spatstat.sparse_3.0-3<br>
[59] httr_1.4.7 polyclip_1.10-6<br>
[61] abind_1.4-5 compiler_4.3.1<br>
[63] withr_3.0.0 BiocParallel_1.34.2<br>
[65] fastDummies_1.7.3 highr_0.10<br>
[67] MASS_7.3-60 DelayedArray_0.26.7<br>
[69] proxyC_0.4.1 tools_4.3.1<br>
[71] lmtest_0.9-40 httpuv_1.6.15<br>
[73] future.apply_1.11.2 goftest_1.2-3<br>
[75] glue_1.6.2 nlme_3.1-162<br>
[77] promises_1.3.0 grid_4.3.1<br>
[79] Rtsne_0.17 cluster_2.1.4<br>
[81] reshape2_1.4.4 generics_0.1.3<br>
[83] gtable_0.3.5 spatstat.data_3.0-4<br>
[85] tidyr_1.3.1 data.table_1.14.8<br>
[87] XVector_0.40.0 sp_2.1-3<br>
[89] utf8_1.2.4 spatstat.geom_3.2-9<br>
[91] RcppAnnoy_0.0.22 ggrepel_0.9.4<br>
[93] RANN_2.6.1 pillar_1.9.0<br>
[95] stringr_1.5.1 limma_3.56.2<br>
[97] spam_2.10-0 RcppHNSW_0.6.0<br>
[99] later_1.3.2 splines_4.3.1<br>
[101] lattice_0.21-8 survival_3.5-5<br>
[103] deldir_2.0-4 tidyselect_1.2.1<br>
[105] locfit_1.5-9.9 miniUI_0.1.1.1<br>
[107] pbapply_1.7-2 gridExtra_2.3<br>
[109] IRanges_2.34.1 SummarizedExperiment_1.30.2 [111] scattermore_1.2
stats4_4.3.1<br>
[113] xfun_0.43 SimBu_1.2.0<br>
[115] matrixStats_1.0.0 stringi_1.7.12<br>
[117] lazyeval_0.2.2 yaml_2.3.8<br>
[119] evaluate_0.23 codetools_0.2-19<br>
[121] SECRET_0.0.0.9000 tibble_3.2.1<br>
[123] cli_3.6.1 uwot_0.2.2<br>
[125] xtable_1.8-4 reticulate_1.36.1<br>
[127] systemfonts_1.0.6 munsell_0.5.1<br>
[129] jquerylib_0.1.4 GenomeInfoDb_1.36.4<br>
[131] Rcpp_1.0.11 globals_0.16.3<br>
[133] spatstat.random_3.2-3 png_0.1-8<br>
[135] parallel_4.3.1 pkgdown_2.0.9<br>
[137] dotCall64_1.1-1 bitops_1.0-7<br>
[139] sparseMatrixStats_1.12.2 listenv_0.9.1<br>
[141] viridisLite_0.4.2 scales_1.3.0<br>
[143] ggridges_0.5.6 crayon_1.5.2<br>
[145] SeuratObject_5.0.1 leiden_0.4.3.1<br>
[147] purrr_1.0.2 rlang_1.1.1<br>
[149] cowplot_1.1.3</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Shanshan Zhang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
